subsys/bluetooth/mesh/crypto_mtls.c

```diff
@@ -0,0 +1,644 @@
+	psa_set_key_algorithm(&attributes, PSA_ALG_CMAC);
+	psa_set_key_type(&attributes, PSA_KEY_TYPE_AES);
+	psa_set_key_bits(&attributes, 128);
+
+	if (psa_import_key(&attributes, key, 16, &key_id) != PSA_SUCCESS) {
+		err = -EIO;
+		goto end;
+	}
+
+	psa_reset_key_attributes(&attributes);
+
+	if (psa_mac_sign_setup(&operation, key_id, alg) != PSA_SUCCESS) {
+		err = -EIO;
+		goto end;
+	}
+
+	for (; sg_len; sg_len--, sg++) {
+		if (psa_mac_update(&operation, sg->data, sg->len) != PSA_SUCCESS) {
+			err = -EIO;
+			goto end;
+		}
+	}
+
+	if (PSA_MAC_LENGTH(PSA_KEY_TYPE_AES, 128, alg) > 16) {
+		err = -ERANGE;
+		goto end;
+	}
+
+	size_t mac_len;
+
+	if (psa_mac_sign_finish(&operation, mac, 16, &mac_len) != PSA_SUCCESS) {
+		err = -EIO;
+		goto end;
+	}
+
+	if (mac_len != 16) {
+		err = -ERANGE;
+	}
+
+end:
+	/* Destroy the key */
+	psa_destroy_key(key_id);
+
+	return err;
+}
+
+static int bt_mesh_aes_cmac_one(const uint8_t key[16], const void *m,
+				size_t len, uint8_t mac[16])
+{
+	struct bt_mesh_sg sg = { m, len };
+
+	return bt_mesh_aes_cmac(key, &sg, 1, mac);
+}
+
+int bt_mesh_s1(const char *m, uint8_t salt[16])
+{
+	const uint8_t zero[16] = { 0 };
+
+	return bt_mesh_aes_cmac_one(zero, m, strlen(m), salt);
+}
+
+int bt_mesh_k1(const uint8_t *ikm, size_t ikm_len, const uint8_t salt[16],
+	       const char *info, uint8_t okm[16])
+{
+	int err;
+
+	err = bt_mesh_aes_cmac_one(salt, ikm, ikm_len, okm);
+	if (err < 0) {
+		return err;
+	}
+
+	return bt_mesh_aes_cmac_one(okm, info, strlen(info), okm);
+}
+
+int bt_mesh_k2(const uint8_t n[16], const uint8_t *p, size_t p_len,
+	       uint8_t net_id[1], uint8_t enc_key[16], uint8_t priv_key[16])
+{
+	struct bt_mesh_sg sg[3];
+	uint8_t salt[16];
+	uint8_t out[16];
+	uint8_t t[16];
+	uint8_t pad;
+	int err;
+
+	BT_DBG("n %s", bt_hex(n, 16));
+	BT_DBG("p %s", bt_hex(p, p_len));
+
+	err = bt_mesh_s1("smk2", salt);
+	if (err) {
+		return err;
+	}
+
+	err = bt_mesh_aes_cmac_one(salt, n, 16, t);
+	if (err) {
+		return err;
+	}
+
+	pad = 0x01;
+
+	sg[0].data = NULL;
+	sg[0].len  = 0;
+	sg[1].data = p;
+	sg[1].len  = p_len;
+	sg[2].data = &pad;
+	sg[2].len  = sizeof(pad);
+
+	err = bt_mesh_aes_cmac(t, sg, ARRAY_SIZE(sg), out);
+	if (err) {
+		return err;
+	}
+
+	net_id[0] = out[15] & 0x7f;
+
+	sg[0].data = out;
+	sg[0].len  = sizeof(out);
+	pad = 0x02;
+
+	err = bt_mesh_aes_cmac(t, sg, ARRAY_SIZE(sg), out);
+	if (err) {
+		return err;
+	}
+
+	memcpy(enc_key, out, 16);
+
+	pad = 0x03;
+
+	err = bt_mesh_aes_cmac(t, sg, ARRAY_SIZE(sg), out);
+	if (err) {
+		return err;
+	}
+
+	memcpy(priv_key, out, 16);
+
+	BT_DBG("NID 0x%02x enc_key %s", net_id[0], bt_hex(enc_key, 16));
+	BT_DBG("priv_key %s", bt_hex(priv_key, 16));
+
+	return 0;
+}
+
+int bt_mesh_k3(const uint8_t n[16], uint8_t out[8])
+{
+	uint8_t id64[] = { 'i', 'd', '6', '4', 0x01 };
+	uint8_t tmp[16];
+	uint8_t t[16];
+	int err;
+
+	err = bt_mesh_s1("smk3", tmp);
+	if (err) {
+		return err;
+	}
+
+	err = bt_mesh_aes_cmac_one(tmp, n, 16, t);
```

Comment:
Why not sizeof(n) instead of 16? More of them bellow.

Reply:
because this is 16 bytes crypto. No possible other digits in the crypto mesh toolchain. Any case this is an initial implementation. I did refactoring related to hardcoded sizes in Mesh 1.1 repo for EHA. No reason to do it here since it will be replaced after merge Mesh 1.1